### ISSUE TITLE
Request to add Modbus Transaction ID

### ISSUE THREAD

**Issue by meeas** [cisagov/icsnpp-modbus#1]
# 💡 Summary #

Merge requests and responses into a single line (most effort) or just add Modbus Transaction ID to outputs (less effort)

## Motivation and context ##

Amazing script guys!  Anyone that can wrangle zeek scripts is at another level!  I love the fact that you are parsing both requests and responses, however it is hard to tell which requests belong to which responses when they are not sequential, which for me is most of the time.

## Implementation notes ##

In my perfect world, I would love to see both requests and responses merged into a single line, which would provide more context and cut the size of the modbus_detailed.log file to about 50% of its current size for the same traffic.  This would be a bit more coding effort to pull off, but I think provide the most value.

Your current output includes:

 - ts
 - uid
 - id.orig_h
 - id.orig_p
 - id.resp_h
 - id.resp_p
 - unit_id
 - func
 - network_direction
 - address
 - quantity
 - values

The problem currently is:

 1. There is no way to associate a request with its response, short of assuming they are sequential, which they often are not, especially when modbus systems use multiple modbus payloads per TCP packet.
 2. Most of those fields are repeated for both request and response, however the address field only appears in requests, and in the case of reads, the values only appear in the response, making it hard to map the two without transaction IDs.

My prefered solution would be to merge requests and responses into a single line with a table format like

 - ts   (timestamp of original request, unless implementation requires this to be response)
 - uid
 - id.orig_h
 - id.orig_p
 - id.resp_h
 - id.resp_p
 - response_ts   (new field showing timestamp of req or resp depending on use in ts field)
 - unit_id
 - transaction_id  (new field)
 - func
 - network_direction    (remove this field as its no longer needed)
 - address
 - quantity
 - values

However if that is too difficult to do in zeek, or if there is some other reason both requests and responses are needed separately, then simply adding the transaction_id field would be sufficient to address the problems.

**Comment by meeas** [comment]
Was thinking, if you can use the first timestamp from the request, having a respond time instead of a timestamp of respond would be more useful.

**Comment by mmguero** [comment]
See https://github.com/zeek/zeek/commit/bab2036aa4fa436d0d237beb4ca2edebac9a8dcc for that commit in Zeek v5.1x

**Comment by mmguero** [comment]
This is icss-modbus running with Zeek v5.1, note the `tid` transaction ID in the log file.

```
podman run -t -i --rm --entrypoint=/bin/bash -v "$(pwd):/$(basename $(pwd)):rw" ghcr.io/mmguero/zeek:latest
Trying to pull ghcr.io/mmguero/zeek:latest...
Getting image source signatures
Storing signatures
root@a946cedb5080:/zeek-logs# cd /icsnpp-modbus/
root@a946cedb5080:/icsnpp-modbus# zkg install .
The following packages will be INSTALLED:
  /icsnpp-modbus (main)

Proceed? [Y/n] y
Running unit tests for "/icsnpp-modbus"
Installing "/icsnpp-modbus"
Installed "/icsnpp-modbus" (main)
Loaded "/icsnpp-modbus"

...

root@a946cedb5080:/tmp/logs# zeek -C -r /icsnpp-modbus/tests/traces/modbus_example.pcap local
WARNING: No Site::local_nets have been defined.  It's usually a good idea to define your local networks.
root@a946cedb5080:/tmp/logs# ls -l
total 80
-rw-r--r-- 1 root root   322 Feb  2 17:04 capture_loss.log
-rw-r--r-- 1 root root   740 Feb  2 17:04 conn.log
-rw-r--r-- 1 root root 30579 Feb  2 17:04 loaded_scripts.log
-rw-r--r-- 1 root root  5058 Feb  2 17:04 modbus.log
-rw-r--r-- 1 root root  5319 Feb  2 17:04 modbus_detailed.log
-rw-r--r-- 1 root root   785 Feb  2 17:04 modbus_mask_write_register.log
-rw-r--r-- 1 root root   938 Feb  2 17:04 modbus_read_write_multiple_registers.log
-rw-r--r-- 1 root root   780 Feb  2 17:04 notice.log
-rw-r--r-- 1 root root   254 Feb  2 17:04 packet_filter.log
-rw-r--r-- 1 root root   867 Feb  2 17:04 stats.log
-rw-r--r-- 1 root root  2885 Feb  2 17:04 telemetry.log
root@a946cedb5080:/tmp/logs# head modbus.log 
#separator \x09
#set_separator	,
#empty_field	(empty)
#unset_field	-
#path	modbus
#open	2023-02-02-17-04-40
#fields	ts	uid	id.orig_h	id.orig_p	id.resp_h	id.resp_p	tid	unit	func	pdu_type	exception
#types	time	string	addr	port	addr	port	count	count	string	string	string
1595660608.237928	CNYovB3f3pajrqccgk	127.0.0.1	47785	127.0.0.1	502	1	4	READ_COILS	REQ	-
1595660608.238144	CNYovB3f3pajrqccgk	127.0.0.1	47785	127.0.0.1	502	1	4	READ_COILS	RESP	-
root@a946cedb5080:/tmp/logs# 
```

**Comment by mmguero** [comment]
Marking as "closed", because as illustrated above, Zeek v5.1's modbus parser includes the transaction ID as part of the modbus.log
